script/verification/HoloceneSystemConfigUpgrade.s.sol

```diff
@@ -0,0 +1,187 @@
+        } else if (sysCfg.version().eq("1.12.0")) {
+            // Supported initial version
+            targetDGF = address(0);
+        } else {
+            revert("unsupported SystemConfig version");
+        }
+    }
+
+    /// @notice Public function that must be called by the verification script.
+    function checkSystemConfigUpgrade() public view {
+        checkTargetVersion();
+        checkScalar();
+        checkDGF();
+        checkGasPayingToken();
+        checkBaseSysCfgVars();
+    }
+
+    function getCodeExceptions() public view returns (address[] memory exceptions) {
+        uint256 len = block.chainid == 1 ? 3 : 4; // Mainnet doesn't need owner exception.
+        exceptions = new address[](len);
+        uint256 i = 0;
+        if (block.chainid != 1) exceptions[i++] = previous.owner;
+        exceptions[i++] = address(uint160(uint256((previous.batcherHash))));
+        exceptions[i++] = previous.unsafeBlockSigner;
+        exceptions[i++] = previous.batchInbox;
+    }
+
+    // Checks semver of SystemConfig is correct after the upgrade
+    function checkTargetVersion() internal view {
+        require(
+            keccak256(abi.encode(getSysCfgVersion())) == keccak256(abi.encode(targetVersion)),
+            "system-config-050: targetVersion"
+        );
+        console.log("confirmed SystemConfig upgraded to version", targetVersion);
+    }
+
+    // Checks scalar, basefeeScalar, blobbasefeeScalar are set consistently:
+    function checkScalar() internal view {
+        uint256 reencodedScalar =
+            (uint256(0x01) << 248) | (uint256(sysCfg.blobbasefeeScalar()) << 32) | sysCfg.basefeeScalar();
+        console.log(
+            "checking baseFeeScalar and blobbaseFeeScalar ",
+            LibString.toString(sysCfg.basefeeScalar()),
+            LibString.toString(sysCfg.blobbasefeeScalar())
+        );
+        if (
+            // If the scalar version (i.e. the most significant bit of the scalar)
+            // is 1, we expect it to be unchanged during the upgrade.
+            uint8(previousScalar >> 248) == 1
+        ) {
+            require(reencodedScalar == previousScalar, "scalar-100 scalar mismatch");
+        } else {
+            // Otherwise, if the scalar version is 0,
+            // and the blobbasefeeScalar is 0,
+            // the upgrade will migrate the scalar version to 1 and preserve
+            // everything else.
+            // See https://specs.optimism.io/protocol/system-config.html?highlight=ecotone%20scalar#ecotone-scalar-overhead-uint256uint256-change
+            require(previousScalar >> 248 == 0, "scalar-101 previous scalar version != 0 or 1");
+            require(reencodedScalar >> 248 == 1, "scalar-102 reenconded scalar version != 1");
+            require(sysCfg.blobbasefeeScalar() == uint32(0), "scalar-103 blobbasefeeScalar !=0");
+            require(reencodedScalar << 8 == previousScalar << 8, "scalar-104 scalar mismatch");
```

Comment:
Two comments on this: 
1. It's not very clear, if we define a variable or at least put in the revert reason "`baseFeeScalar`" that would help a lot
2. The bit shift looks wrong, we probably want  `<< 32` instead of `<< 8`

Reply:
Looks like this is not working indeed, i tried with the mask `0xfffff` and this seems to have a more reasonable value @geoknee: 
![14aa2bca398d7e38779daed1c1b54d44fb9d852fb8378e0b4641ebcf9c47b124](https://github.com/user-attachments/assets/959348b9-a03d-4a7e-9eaf-2bafb8daf00c)

Reply:
Fixed this here. In fact what I was trying to do was compare the scalars _without_ the scalar version (first byte). [556d1d5](https://github.com/ethereum-optimism/superchain-ops/pull/418/commits/556d1d51bc7b73b13cc9aeda26cb54c042f589ad)

Reply:
Hmm, 
Just to make sure, you want to compare the if the first byte (`0x0`) are correctly incremented to (`0x1`) when the value `scalarVersion` is `0` right (if yes for me it works as expected). 
![CleanShot 2025-01-10 at 13 05 23@2x](https://github.com/user-attachments/assets/0912c0e9-1cf1-470d-b3f5-e032e938c91e)

Reply:
Yes that's right.